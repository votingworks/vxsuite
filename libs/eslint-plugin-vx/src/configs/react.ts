import recommended = require('./recommended');

const jsExtensions = ['.js', '.jsx'];
const tsExtensions = ['.ts', '.tsx'];
const allExtensions = jsExtensions.concat(tsExtensions);

export = {
  ...recommended,
  extends: [
    ...recommended.extends
      .map((name) => (name === 'airbnb-base' ? 'airbnb' : name))
      .filter((name) => name !== 'plugin:prettier/recommended'),
    'airbnb/hooks',
    'plugin:react/recommended',
    'plugin:jsx-a11y/recommended',
    'plugin:prettier/recommended', // Enables eslint-plugin-prettier and displays prettier errors as ESLint errors. Make sure this is always the last configuration in the extends array.
  ],
  settings: {
    ...recommended.settings,
    react: {
      version: 'detect',
    },
  },
  rules: {
    ...recommended.rules,
    'jsx-a11y/label-has-associated-control': [
      'error',
      {
        controlComponents: ['Select'],
      },
    ],
    'react/jsx-filename-extension': ['error', { extensions: allExtensions }],
    'react/jsx-fragments': ['error', 'element'],
    'react/jsx-no-bind': 'off',
    'react/jsx-no-constructed-context-values': 'off',
    'react/jsx-no-useless-fragment': ['error', { allowExpressions: true }],
    'react/jsx-one-expression-per-line': 'off',
    'react/jsx-props-no-spreading': 'off',
    'react/jsx-wrap-multilines': 'off',
    'react/prop-types': 'off',
    'react/require-default-props': 'off',
  },
};
