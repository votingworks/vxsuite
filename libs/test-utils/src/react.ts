/* eslint-disable no-console -- this file monkey-patches `console` */

/**
 * Suppresses the "Can't perform a React state update on an unmounted component" warning by
 * modifying the `console.error` method to ignore the warning.
 *
 * This is a temporary workaround until we upgrade to React 18.
 */
export function suppressReact17UnmountedWarning(): void {
  const originalConsoleError = console.error;
  console.error = (...args: unknown[]) => {
    if (
      typeof args[0] === 'string' &&
      args[0].startsWith(
        `Warning: Can't perform a React state update on an unmounted component`
      )
    ) {
      return;
    }

    originalConsoleError(...args);
  };
}
