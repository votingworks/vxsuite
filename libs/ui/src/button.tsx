import React, { useState } from 'react';
import styled, { css, StyledComponent } from 'styled-components';
import { EventTargetFunction } from '@votingworks/types';

export interface ButtonInterface {
  readonly danger?: boolean;
  readonly fullWidth?: boolean;
  readonly large?: boolean;
  readonly noFocus?: boolean;
  readonly noWrap?: boolean;
  readonly primary?: boolean;
  readonly small?: boolean;
  readonly textAlign?: 'left' | 'center' | 'right';
  readonly warning?: boolean;
}

export interface StyledButtonProps
  extends ButtonInterface,
    React.PropsWithoutRef<JSX.IntrinsicElements['button']> {}

export const buttonNoFocusOutlineStyle = css`
  outline: none;
`;

const buttonStyles = css<StyledButtonProps>`
  display: inline-block;
  border: none;
  border-radius: 0.25em;
  box-shadow: 0 0 0 0 rgba(71, 167, 75, 1);
  box-sizing: border-box;
  background: ${({ disabled, danger, warning, primary }) =>
    (disabled && 'rgb(211, 211, 211)') ||
    (danger && 'red') ||
    (warning && 'darkorange') ||
    (primary && 'rgb(71, 167, 75)') ||
    'rgb(211, 211, 211)'};
  cursor: ${({ disabled = false }) => (disabled ? undefined : 'pointer')};
  width: ${({ fullWidth = false }) => (fullWidth ? '100%' : undefined)};
  padding: ${({ large = false, small = false }) =>
    small ? '0.35em 0.5em' : large ? '1em 1.75em' : '0.75em 1em'};
  text-align: ${({ textAlign }) => textAlign || 'center'};
  line-height: 1.25;
  color: ${({ disabled, danger, warning, primary }) =>
    (disabled && 'rgb(160, 160, 160)') ||
    (danger && '#FFFFFF') ||
    (warning && '#FFFFFF') ||
    (primary && '#FFFFFF') ||
    'black'};
  font-size: ${({ large = false }) => (large ? '1.25em' : undefined)};
  touch-action: manipulation;
  &:focus {
    /* stylelint-disable-next-line value-keyword-case */
    ${({ noFocus = false }) =>
      noFocus ? buttonNoFocusOutlineStyle : undefined}
  }
  &:hover,
  &:active {
    outline: none;
  }
`;

export const DecoyButton = styled.div`
  ${buttonStyles}/* stylelint-disable-line value-keyword-case */
`;

const StyledButton = styled('button').attrs(({ type = 'button' }) => ({
  type,
}))`
  ${buttonStyles}/* stylelint-disable-line value-keyword-case */
`;

export interface ButtonProps extends StyledButtonProps {
  component?: StyledComponent<'button', never, StyledButtonProps, never>;
  onPress: EventTargetFunction;
  ref?: React.Ref<HTMLButtonElement>;
}

export const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  (
    { component: Component = StyledButton, onPress, disabled, ...rest },
    ref
  ) => {
    const [startCoordinates, setStartCoordinates] = useState([0, 0]);

    function onTouchStart(event: React.TouchEvent) {
      const { clientX, clientY } = event.touches[0];
      setStartCoordinates([clientX, clientY]);
    }

    function onTouchEnd(event: React.TouchEvent) {
      const maxMove = 30;
      const { clientX, clientY } = event.changedTouches[0];
      if (
        !disabled &&
        Math.abs(startCoordinates[0] - clientX) < maxMove &&
        Math.abs(startCoordinates[1] - clientY) < maxMove
      ) {
        onPress(event);
        event.preventDefault();
      }
    }

    return (
      <Component
        {...rest}
        disabled={disabled}
        onTouchStart={onTouchStart}
        onTouchEnd={onTouchEnd}
        onClick={onPress}
        ref={ref}
      />
    );
  }
);

Button.displayName = 'Button';

export const SegmentedButton = styled.span`
  display: inline-flex;
  white-space: nowrap;
  & > button {
    box-shadow: inset 1px 0 0 rgb(190, 190, 190);
  }
  & > button:first-child {
    box-shadow: none;
  }
  & > button:not(:last-child) {
    border-top-right-radius: 0;
    border-bottom-right-radius: 0;
  }
  & > button:not(:first-child) {
    border-top-left-radius: 0;
    border-bottom-left-radius: 0;
  }
  & > button:disabled {
    background: #028099;
    color: #ffffff;
  }
`;

export const LabelButton = styled.label`
  ${buttonStyles}/* stylelint-disable-line value-keyword-case */
`;
