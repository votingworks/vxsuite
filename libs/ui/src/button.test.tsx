import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Button, DecoyButton } from './button';

function createTouchStartEventProperties(x: number, y: number) {
  return { touches: [{ clientX: x, clientY: y }] };
}

function createTouchEndEventProperties(x: number, y: number) {
  return { changedTouches: [{ clientX: x, clientY: y }] };
}

describe('renders Button', () => {
  test('with defaults', () => {
    const { container } = render(<Button onPress={jest.fn()}>foo</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('with options: primary noFocus', () => {
    const { container } = render(
      <Button onPress={jest.fn()} primary noFocus>
        Primary
      </Button>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test('with options: full-width disabled submit', () => {
    const { container } = render(
      <Button onPress={jest.fn()} type="submit" disabled fullWidth>
        Disabled Button
      </Button>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test('with options: big danger', () => {
    const { container } = render(
      <Button onPress={jest.fn()} large danger>
        Iâ€™m a big button!
      </Button>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test('as DecoyButton with options: small warning', () => {
    const { container } = render(
      <DecoyButton small warning>
        DecoyButton
      </DecoyButton>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test('and tests clicks and touches', () => {
    const onPress = jest.fn();
    const { getByText } = render(
      <Button onPress={onPress}>Test Button</Button>
    );
    const button = getByText('Test Button');

    fireEvent.click(button);
    expect(onPress).toHaveBeenCalledTimes(1);

    // TouchEnd close to TouchStart calls onPress.
    fireEvent.touchStart(button, createTouchStartEventProperties(100, 100));
    fireEvent.touchEnd(button, createTouchEndEventProperties(110, 95));
    expect(onPress).toHaveBeenCalledTimes(2);

    // Using preventDefault() with touch prevents the click, so no need to test click de-duping.

    // TouchStart w/o TouchEnd does not call onPress.
    fireEvent.touchStart(button, createTouchStartEventProperties(100, 100));
    expect(onPress).toHaveBeenCalledTimes(2);

    // TouchEnd too far from TouchStart does not call onPress.
    fireEvent.touchEnd(button, createTouchEndEventProperties(131, 95));
    expect(onPress).toHaveBeenCalledTimes(2);

    // Keyboard (also Accessible Controller) fire click event which calls onPress.
    fireEvent.click(button);
    expect(onPress).toHaveBeenCalledTimes(3);
  });
});

test('disabling works for both touch events and click events', () => {
  const onPress = jest.fn();
  render(
    <Button onPress={onPress} disabled>
      Disabled Button
    </Button>
  );
  const button = screen.getByText('Disabled Button');

  // Click is disabled
  userEvent.click(button);
  expect(onPress).toHaveBeenCalledTimes(0);

  // Touch is disabled
  fireEvent.touchStart(button, createTouchStartEventProperties(100, 100));
  fireEvent.touchEnd(button, createTouchEndEventProperties(100, 100));
  expect(onPress).toHaveBeenCalledTimes(0);
});
