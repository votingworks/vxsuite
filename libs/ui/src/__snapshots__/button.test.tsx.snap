// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Button as DecoyButton with options: small warning 1`] = `
.c0 {
  display: inline-block;
  border: none;
  border-radius: 0.25em;
  box-shadow: 0 0 0 0 rgba(71,167,75,1);
  box-sizing: border-box;
  background: darkorange;
  cursor: pointer;
  padding: 0.35em 0.5em;
  text-align: center;
  line-height: 1.25;
  color: #FFFFFF;
  touch-action: manipulation;
}

.c0:hover,
.c0:active {
  outline: none;
}

<div
  class="c0"
>
  DecoyButton
</div>
`;

exports[`renders Button with options: big danger 1`] = `
.c0 {
  display: inline-block;
  border: none;
  border-radius: 0.25em;
  box-shadow: 0 0 0 0 rgba(71,167,75,1);
  box-sizing: border-box;
  background: red;
  cursor: pointer;
  padding: 1em 1.75em;
  text-align: center;
  line-height: 1.25;
  color: #FFFFFF;
  font-size: 1.25em;
  touch-action: manipulation;
}

.c0:hover,
.c0:active {
  outline: none;
}

<button
  class="c0"
  type="button"
>
  Iâ€™m a big button!
</button>
`;

exports[`renders Button with options: full-width disabled submit 1`] = `
.c0 {
  display: inline-block;
  border: none;
  border-radius: 0.25em;
  box-shadow: 0 0 0 0 rgba(71,167,75,1);
  box-sizing: border-box;
  background: rgb(211,211,211);
  width: 100%;
  padding: 0.75em 1em;
  text-align: center;
  line-height: 1.25;
  color: rgb(160,160,160);
  touch-action: manipulation;
}

.c0:hover,
.c0:active {
  outline: none;
}

<button
  class="c0"
  disabled=""
  type="submit"
>
  Disabled Button
</button>
`;

exports[`renders Button with options: primary noFocus 1`] = `
.c0 {
  display: inline-block;
  border: none;
  border-radius: 0.25em;
  box-shadow: 0 0 0 0 rgba(71,167,75,1);
  box-sizing: border-box;
  background: rgb(71,167,75);
  cursor: pointer;
  padding: 0.75em 1em;
  text-align: center;
  line-height: 1.25;
  color: #FFFFFF;
  touch-action: manipulation;
}

.c0:focus {
  outline: none;
}

.c0:hover,
.c0:active {
  outline: none;
}

<button
  class="c0"
  type="button"
>
  Primary
</button>
`;
